### PR TITLE
Declaration-emit class expressions as type literals

### LINKED ISSUE #15066: Using mixins throws TS4093, TS4020 and TS4060 when exporting the class that extends the result of a mixin
**TypeScript Version:**  2.2.1

**Code**

I'm having some problems with using the new syntax for mixins. Consider the following example:
```ts
export abstract class Item {
    foo(): void {}
}

export class FooItem extends Item {
	name?: string;

	toString(): string {
		if (this.name) {
			return this.name;
		}
		return `${this.id}`;
	}
}


export type Constructor<T> = new(...args: any[]) => T;

// Throws an error when compiled
// "Error TS4060: Return type of exported function has or is using private name '(Anonymous class)'.
export function WithTags<T extends Constructor<FooItem>>(Base: T) {
	return class extends Base {
		static getTags(): Promise<any> { ... }
		tags(): Promise<any> { ... }
	}
}

// Throws two errors:
// At `Test`: "Error TS4093: 'extends' clause of exported class 'Test' refers to a type whose name cannot be referenced."
// At `WithTags`: "Error TS4020: 'extends' clause of exported class 'Test' has or is using private name '(Anonymous class)'."
export class Test extends WithTags(FooItem) {}

const test = new Test();

Test.getTags()
test.tags();
```

**Expected behavior:**
It should not throw any errors.

**Actual behavior:**
It throws those three errors that are described in the code comments.

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #15066, although the resulting d.ts file is a bit ugly. After discussion, @mhegazy and I decided that adding a new "named class expression" type would not be enough to solve the problem on its own, making any complete solution extremely complex. Instead, this change just writes out class expressions as object types, resulting in types like this:

```ts
// @Filename: a.ts
class B { b; }
function f() {
    return class extends B { static d; }
}
// @Filename: a.d.ts
declare class B { b: any; }
declare function f(): {
  new(): { b: any };
  prototype: { b: any };
  d: any;
}
```

Notes:
1. Circular references bottom out as `any`.
2. This happens for all type writing, not just declaration emit, but this is probably a slightly improvement overall.

### CODE REVIEW

File: tests/baselines/reference/emitClassExpressionInDeclarationFile.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+Test.getTags();
+test.tags();
+
+
+//// [emitClassExpressionInDeclarationFile.d.ts]
+export declare var simpleExample: {
+    new (): {
+        tags(): void;
+    };
+    prototype: {
```
Comment (mhegazy) [inline tests/baselines/reference/emitClassExpressionInDeclarationFile.js]: kinda weird that we keep this property. can we filter `prototype` out?

File: tests/cases/fourslash/findAllRefsForVariableInExtendsClause01.ts [hunk 2]
```diff
@@ -3,4 +3,4 @@
 ////var [|{| "isWriteAccess": true, "isDefinition": true |}Base|] = class { };
 ////class C extends [|Base|] { }
 
-verify.singleReferenceGroup("var Base: typeof Base");
+verify.singleReferenceGroup("var Base: {\n    new (): {};\n    prototype: {};\n}");
```
Comment (sheetalkamat) [inline tests/cases/fourslash/findAllRefsForVariableInExtendsClause01.ts]: If there is a way to keep this as typeof Base it would be better I think. The completionlist/quickinfoes would get unreadable with this change otherwise

File: tests/cases/compiler/emitClassExpressionInDeclarationFile.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+export type Constructor<T> = new(...args: any[]) => T;
+export function WithTags<T extends Constructor<FooItem>>(Base: T) {
+    return class extends Base {
+        static getTags(): void { }
+        tags(): void { }
+    }
+}
+
+export class Test extends WithTags(FooItem) {}
+
```
Comment (sheetalkamat) [inline tests/cases/compiler/emitClassExpressionInDeclarationFile.ts]: can you please add example like:
```typescript
class Base {
}
export function foo() {
return class extends Base {
}
}
```

This should report error that Base is private and used here.

File: tests/cases/compiler/emitClassExpressionInDeclarationFile.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+// @declaration: true
+export var simpleExample = class {
+    static getTags() { }
+    tags() { }
```
Comment (sheetalkamat) [inline tests/cases/compiler/emitClassExpressionInDeclarationFile.ts]: Can you also add private member to this? Are we suppose to see private member in the type literal? If yes we should not be writing type of that private property/method because then it would likely start showing errors (with inaccessible symbol)

File: tests/cases/fourslash/findAllRefsForVariableInExtendsClause01.ts (hunk 2 above)
Comment (sandersn) [inline tests/cases/fourslash/findAllRefsForVariableInExtendsClause01.ts]: Done

File: tests/baselines/reference/emitClassExpressionInDeclarationFile.js (hunk 1 above)
Comment (sandersn) [inline tests/baselines/reference/emitClassExpressionInDeclarationFile.js]: Done

File: tests/cases/compiler/emitClassExpressionInDeclarationFile.ts (hunk 4 above)
Comment (sandersn) [inline tests/cases/compiler/emitClassExpressionInDeclarationFile.ts]: Added test, and removed privates from declaration emit.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -3459,6 +3479,11 @@ namespace ts {
                     buildIndexSignatureDisplay(resolved.stringIndexInfo, writer, IndexKind.String, enclosingDeclaration, globalFlags, symbolStack);
                     buildIndexSignatureDisplay(resolved.numberIndexInfo, writer, IndexKind.Number, enclosingDeclaration, globalFlags, symbolStack);
                     for (const p of resolved.properties) {
+                        if (globalFlags & TypeFormatFlags.WriteClassExpressionAsTypeLiteral &&
+                            (p.name === "prototype" ||
```
Comment (sheetalkamat) [inline src/compiler/checker.ts]: how about instead checking ```SymbolFlags.Prototype```

Comment (sandersn) [inline src/compiler/checker.ts]: Good idea.

File: tests/cases/compiler/emitClassExpressionInDeclarationFile.ts (hunk 3 above)
Comment (sandersn) [inline tests/cases/compiler/emitClassExpressionInDeclarationFile.ts]: I tried adding an ad-hoc check, but I couldn't get it to distinguish between this case, where the class expression extends a class declaration, and the mixin case, where the class expression extends a parameter. There are two problems with the latter case:

1. The mixed-in class could easily be a non-exported class declaration.
2. Class declarations that call the mix-in are no longer enclosed in the mix-in function, making it difficult to track back to the original class expression's context in the displayType machinery.

This export will not cause errors in any case, because the class expression is exported as a type literal, so `Base` isn't actually referenced &mdash; just its contents.

tl;dr &mdash; I don't think it's worthwhile to issue an error here.

**Comment by sandersn** [comment]
All right, I think this change is ready to go in now. @mhegazy mind looking at the latest changes to emit errors for private properties that are part of exported class expressions?

**Comment by mihailik** [comment]
How about using function-aliased namespace?

```typescript
// @Filename: a.ts
class B { b; }
function f() {
    return class extends B { static d; }
}
// @Filename: a.d.ts
declare class B { b: any; }
declare function f(): B._return;
declare namespace f {
  class _return extends B  { static d; }
}
```

As you've said, it's not solving the generic case. But it's a help of a lot easier to read!

**Comment by jtlapp** [comment]
Is it possible to apply this solution to mixins that use generics? I have the following mixins working, as long as I don't try to use them from within another module:

```ts
// broker.ts

export class BasicBroker<E> {
    doSomethingBasic(): E {
        //...
    }
}

export function SpecialBrokerMixin<E>() {
    return function <B extends Constructor<{}>>(Base: B) {
        // allows mixins to use decorators and be declared abstract
        class WithSpecialBroker extends Base {
            doSomethingSpecial(): E {
                //...
            }
        }
        return WithSpecialBroker;
    }
}

export function BasicSpecialBroker<E>() {
    // Can't pass a generic to a function. so make concrete first.
    class ConcreteBaseBroker extends BasicBroker<E> { }
    return SpecialBrokerMixin<E>()(ConcreteBaseBroker);
}
```

Here's what happens if I try to import them:

```ts
// app.ts

class AppEntity {
    //...
}

// Error TS4020: 'extends' clause of exported class 'AppBroker' has or is using private name 'ConcreteBaseBroker'.
// Error TS4020: 'extends' clause of exported class 'AppBroker' has or is using private name 'WithSpecialBroker'.

export class AppBroker extends BasicSpecialBroker<AppEntity>() {
    //...
}
```

I don't see how to define either `ConcreteBaseBroker` or `WithSpecialBroker` in ambient typings.

**Comment by jtlapp** [comment]
I found the problem. The compiler `--declaration` option appears to break things. Everything works if I turn off that option. Should that be a bug report?
